core/src/Kokkos_CopyViews.hpp

```diff
@@ -1401,7 +1401,33 @@ void deep_copy
   typedef typename src_type::memory_space     src_memory_space ;
   typedef typename dst_type::value_type       dst_value_type ;
   typedef typename src_type::value_type       src_value_type ;
-  if(dst.data() == NULL && src.data() == NULL) {
+  if(dst.data() == NULL || src.data() == NULL) {
```

Comment:
A question not relevant to this fix: Why can't this specialization of `deep_copy` just call the one that takes an execution space instance as its first argument?

Reply:
Good point Mark.  I just went over the source code and I believe calling the other one with the execution space of the destination would do the trick.

Reply:
This would remove ~200 lines of essentially duplicated code. I am trying it locally, building the tests at the moment.  Haven't looked into the other 3 pairs of `deep_copy(dst, src)` and `deep_copy(exec_space, dst, src)` in details to see if there was more opportunity to downsize the amount of code to maintain.

Reply:
This could have a merge conflict with https://github.com/kokkos/kokkos/pull/1919 , but in general is in the spirit of making Kokkos respect execution space instances throughout.

Reply:
@mhoemmen Some tests were failing now.  It is not obvious to me what went wrong, whether I overlooked some subtle difference between the two functions (most likely the issue...) or if I messed up when I implemented the changes you suggested.

Reply:
@dalg24 I remember trying this once before and having the same issue.  Thanks for trying, though!  It's fine to duplicate that code for now.